docs/spec/broker.md

```diff
@@ -118,6 +118,11 @@ Events contained in delivery responses SHOULD be published to the Broker ingress
 and processed as if the event had been produced to the Broker's addressable
 endpoint.
 
+Events contained in delivery responses that are malformed SHOULD be treated as
```

Comment:
Can be fixed in a followup, but this should probably be more specific about how a response event is known to be malformed.

Reply:
Ack, I'm hoping to rope in @slinkydeveloper and @n3wscott here for that. I actually thought that it may belong to Interfaces (in particular Callable). So, if here we say it's malformed, and in the Callable contract we specify the detection, seems cleaner and more applicable to not just Broker? Thoughts?

Reply:
:+1: IMO it should be in the delivery spec at https://github.com/knative/eventing/blob/master/docs/spec/data-plane.md

Reply:
Yes, so in particular this Callable section, might just need to changed to say something that the returned event must be a valid Cloud Event.

```
If Sink is Callable it MAY respond with 200 OK and a single event in the HTTP response. A returned event is not required to be related to the received event. The Callable should return a successful response if the event was processed successfully. If there is no event to send back then Callable Sink MUST respond with 2xx HTTP and with empty body.

```

Reply:
Is it ok that a valid empty non cloudevent response (200/202 without body and without ce headers) is treated as a failure?

As you pointed out:

> If Sink is Callable it MAY respond with 200 OK and a single event in the HTTP response[...] If there is no event to send back then Callable Sink MUST respond with 2xx HTTP and with empty body.

But with this change you're making the first MAY a MUST, because this condition https://github.com/knative/eventing/pull/3450/files#diff-a9cf39e105b4ea1643e25fc890eb8462R268 will catch even a 200/202 empty body and no ce headers. The correct condition to fullfil this contract should be:

```golang
response := cehttp.NewMessageFromHttpResponse(resp)
defer response.Finish(nil)

if response.ReadEncoding() == binding.EncodingUnknown {
        // Response doesn't have a ce-specversion header nor a content-type matching a cloudevent event format
	if response.BodyReader is not empty {
		//no event in response, but body not empty -> fail
	}
	// no event in response and empty body
	return resp.StatusCode, nil
}

event, err := binding.ToEvent(ctx, response)
if err != nil {
	// event in response but something failed (eg wrong ce-specversion)
	return resp.StatusCode, err
}
```

Reply:
I think as a user of the sdk, it would be cleaner to have binding.ToEvent to return nil if there's no event in the response. I'm probably thinking of this wrong, but I would prefer that the sdk would return a nil, nil rather than an nil, error for this case. If there's no event there (no data), there's no event and no error. Just my $.02.

Reply:
That's fair, returning `binding.ErrUnknownEncoding` is a choice I made mostly to be consistent with other signatures of the `binding` package, among the others `binding.Write`. But because this is a breaking change, I'm not sure we can do this change soon in a next minor of the sdk.

In any case, i think you should handle the case _200/202, empty body and no ce headers_ separately :smile:

Reply:
ok, added bunch more code :) I'd like to see about changing the behaviour in the sdk however in the future.